[vpj] Replace numeric offsets with PubSubPosition in VPJ repush

sushantmane opened the pull request:
## [vpj] Replace numeric offsets with PubSubPosition in VPJ repush 

Refactor MR and Spark repush to use PubSubPartitionSplit, SplitRequest,  
PubSubSplitPlanner, and splitting strategies introduced in the previous PR.  
Replace numeric record offset comparator with a new local index comparator to  
avoid PubSubPosition comparisons in contexts where dependencies cannot be  
injected.  

Numeric offsets remain the default comparator to ensure safe ordering across  
records. The local index comparator is gated behind a config flag, providing  
an alternative that eliminates PubSubPosition dependency overhead. This option  
can reduce complexity but carries rare edge cases if log compaction occurs  
during repush and speculative execution or failures cause reprocessing.  

These changes consolidate repush logic across Hadoop MR and Spark, remove  
duplicate code paths, and align record ordering strategies with the new  
abstraction layer for easier extension and maintenance.  

###  Code changes
- [ ] Added new code behind **a config**. If so list the config names and their default values in the PR description.
- [ ] Introduced new **log lines**. 
  - [ ] Confirmed if logs need to be **rate limited** to avoid excessive logging.

###  **Concurrency-Specific Checks**
Both reviewer and PR author to verify
- [ ] Code has **no race conditions** or **thread safety issues**.
- [ ] Proper **synchronization mechanisms** (e.g., `synchronized`, `RWLock`) are used where needed.
- [ ] No **blocking calls** inside critical sections that could lead to deadlocks or performance degradation.
- [ ] Verified **thread-safe collections** are used (e.g., `ConcurrentHashMap`, `CopyOnWriteArrayList`).
- [ ] Validated proper exception handling in multi-threaded code to avoid silent thread termination.


## How was this PR tested?
<!--
If you're unsure about what to test, where to add tests, or how to run tests, please feel free to ask. We'd be happy to help.
-->

- [ ] New unit tests added.
- [ ] New integration tests added.
- [ ] Modified or extended existing tests.
- [ ] Verified backward compatibility (if applicable).

## Does this PR introduce any user-facing or breaking changes?
<!--  
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, choose 'No'.
-->
- [ ] No. You can skip the rest of this section.
- [ ] Yes. Clearly explain the behavior change and its impact.

arjun4084346 commented on internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicMetadataFetcher.java:
```diff
@@ -345,37 +307,91 @@ List<PubSubTopicPartitionInfo> getTopicPartitionInfo(PubSubTopic topic) {
+      long t0 = System.nanoTime();
+      Map<PubSubTopicPartition, PubSubPosition> res =
+          consumer.endPositions(parts, getPubsubOffsetApiTimeoutDurationDefaultValue());
+      stats.recordLatency(GET_TOPIC_LATEST_OFFSETS, t0);
```
it should be END_POSITIONS
or t0 should be initialized at beginning of the call

sushantmane commented on clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java:
```diff
@@ -294,15 +307,26 @@ public void close() {
     // not most accurate but gives reasonable estimate
-    return ((float) (currentOffset - startingOffset + 1)) / maxNumberOfRecords;
+    if (totalRecordsToRead <= 0) {
+      return 0.0f;
```
Good point.

totalRecordsToRead = 0 is unexpected since we don't generate split with zero records to consume.

sushantmane commented on clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java:
```diff
@@ -294,15 +307,26 @@ public void close() {
      */
-    return currentOffset < endingOffset - 1;
+    return recordsReadSoFar.get() < totalRecordsToRead
+        || consumer.positionDifference(topicPartition, endingPosition, currentPosition) > 1;
```
No, it can’t be, because computing the delta between PubSubPositions requires passing the topic handle, which is resolved using PubSubConsumer by making call to the PubSub backend 

We do have a static variant of this method in PubSubUtil, but we’re replacing it with TM/Consumer.positionDifference for the reason above.

sushantmane commented on internal/venice-common/src/main/java/com/linkedin/venice/pubsub/manager/TopicMetadataFetcher.java:
```diff
@@ -345,37 +307,91 @@ List<PubSubTopicPartitionInfo> getTopicPartitionInfo(PubSubTopic topic) {
+      if (parts.isEmpty()) {
+        return Collections.emptyMap();
+      }
+      return consumer.beginningPositions(parts, getPubsubOffsetApiTimeoutDurationDefaultValue());
```
We recently removed most metrics from TopicManagerStats to reduce unnecessary bloat, but kept GET_TOPIC_LATEST_OFFSETS since it has proven useful on a few occasions.

sushantmane commented on clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputDictTrainer.java:
```diff
@@ -182,7 +182,7 @@ synchronized byte[] trainDict(Optional<PubSubConsumerAdapter> reusedConsumerOpti
     // Prepare input
     // Get one split per partition
-    KafkaInputSplit[] splits = (KafkaInputSplit[]) kafkaInputFormat.getSplitsByRecordsPerSplit(jobConf, Long.MAX_VALUE);
+    KafkaInputSplit[] splits = (KafkaInputSplit[]) kafkaInputFormat.getSplits(jobConf, Integer.MAX_VALUE);
```
Thanks for pointing this out. I had accidentally changed the behavior documented in the comment. I've made changes to revert it back
```
// Prepare input
    // Get one split per partition
```

sushantmane commented:
Thanks, @arjun4084346, for the review. I’ve incorporated your comments. Please take another look when you have a chance.

sixpluszero commented on clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/pubsub/input/PubSubPartitionSplit.java:
```diff
@@ -0,0 +1,235 @@
+  }
+
+  private void ensureRepo() {
+    if (topicRepository == null) {
```
Hmm, you already have: "    this.topicRepository = Objects.requireNonNull(topicRepository, "topicRepository");" this seems unnecessary

sushantmane replied:
yeah and we create it in nullary constructor as well. Removed

sixpluszero commented on clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/pubsub/input/splitter/FixedRecordCountRangeSplitter.java:
```diff
@@ -0,0 +1,79 @@
+              curEnd, // exclusive
+              thisSplitCount, // numberOfRecords
+              rangeIndex,
+              startOffset)); // rangeIndex
```
Looks like this one directly use local split index instead of being controlled by feature flag?

sushantmane replied:
splits are assigned local indexes irrespective of whether we use them or not.

sixpluszero reviewed (commented):
Thank you for the change! Finished a pass on the main code change and left some comments. The PR is a little bit too big so I ask some dumb clarifications here to make sure I understand it fully, I didn't get to check most of the tests files but I assume most of the unit test is AI generated :) 
(would be great if we could do two PR in the future for this size of change like one introduce the split and all its types implementation and 2nd one integrating it into the original impl)

sushantmane commented on clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputFormat.java:
```diff
@@ -43,77 +27,32 @@
-   * BTW, this calculation is not accurate since it is purely based on offset, and the topic
-   * being consumed could have log compaction enabled.
-   */
-  public static final long DEFAULT_KAFKA_INPUT_MAX_RECORDS_PER_MAPPER = 5000000L;
```
Those remain unchanged. Previously, these configs were scattered throughout the code, but they’ve now been consolidated into the VPJ constants class.

sushantmane commented on clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java:
```diff
@@ -294,15 +304,26 @@ public void close() {
+      return 1.0f;
+    }
+    float progress = Math.min(1.0f, ((float) recordsReadSoFar.get() / totalRecordsToRead));
+    LOGGER.debug(
```
Updated to log this after each 1% progress. So max log lines per split/mapper will be  100

sushantmane commented on clients/venice-push-job/src/main/java/com/linkedin/venice/spark/input/pubsub/SparkPubSubPartitionReader.java:
```diff
@@ -92,42 +99,56 @@ public VeniceRawPubsubInputPartitionReader(
+      return 1.0f;
+    }
+    float progress = Math.min(1.0f, ((float) recordsReadSoFar.get() / totalRecordsToRead));
+    LOGGER.debug(
```
Changed to log progress at every 1% increment

sushantmane commented on clients/venice-push-job/src/main/java/com/linkedin/venice/spark/input/pubsub/VenicePubSubMessageToRow.java:
```diff
@@ -26,14 +27,9 @@ public class VenicePubSubMessageToRow implements PubSubMessageConverter {
-    buffer.get(bytes); // this stamps the bytes with contents of the buffer
-    return bytes;
+      int partitionNumber,
+      long offset) {
```
It's passed from `SparkPubSubPartitionReader`
```
      this.currentRow = this.pubSubMessageConverter.convert(message, this.region, this.targetPartitionNumber, offset);

```

sushantmane replied:
>I wonder if your intention to change it is because you want the customized local offset to be passed in


That's right

arjun4084346 commented on clients/venice-push-job/src/main/java/com/linkedin/venice/hadoop/input/kafka/KafkaInputRecordReader.java:
```diff
@@ -294,15 +305,32 @@ public void close() {
     // not most accurate but gives reasonable estimate
-    return ((float) (currentOffset - startingOffset + 1)) / maxNumberOfRecords;
+    if (totalRecordsToRead <= 0) {
+      return 1.0f;
```
maybe `Float.NaN` is more appropriate

sushantmane replied:
Not sure that will break MR. Keeping update change from 0 to 1 as suggested in prev comment

arjun4084346 commented:
can this pr be broken into smaller pieces? there are multiple tasks mentioned in the description - 1) offset to position change 2) Introduce per-consumer monotonic 3) Refactor and abstract previously duplicated record reading and split calculation. can these be separated ?

arjun4084346 commented on internal/venice-common/src/main/java/com/linkedin/venice/pubsub/PubSubPositionFactory.java:
```diff
@@ -37,24 +37,16 @@ public PubSubPositionFactory(int positionTypeId) {
    */
-  public PubSubPosition fromPositionRawBytes(PubSubPositionWireFormat positionWireFormat) {
-    if (positionWireFormat.getType() != positionTypeId) {
+  public PubSubPosition fromWireFormat(PubSubPositionWireFormat wireFormat) {
```
api name/signature changes break the dependent projects. one way could be to introduce the new api without removing the old, and annotating as `@ApiStatus.ScheduledForRemoval`

sushantmane replied:
Updated.

arjun4084346 commented on clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/pubsub/input/PubSubPartitionSplit.java:
```diff
@@ -0,0 +1,225 @@
+      PubSubPosition endPubSubPosition,
+      long numberOfRecords,
+      int splitIndex,
+      long startIndex) {
```
it should be `long startOffset`

arjun4084346 commented on clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/pubsub/input/splitter/TimeWindowRangeSplitter.java:
```diff
@@ -0,0 +1,13 @@
+public class TimeWindowRangeSplitter implements PartitionRangeSplitter {
+  @Override
+  public List<PubSubPartitionSplit> split(SplitRequest splitRequest) {
+    throw new UnsupportedOperationException("TimeWindowRangeSplitter is not supported");
```
we have plans to support it in future?

arjun4084346 commented on clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/pubsub/input/PubSubSplitPlanner.java:
```diff
@@ -0,0 +1,111 @@
+
+    for (int p = 0; p < partitionCount; p++) {
+      PubSubTopicPartition tp = new PubSubTopicPartitionImpl(topic, p);
+      SplitRequest req = new SplitRequest.Builder().pubSubTopicPartition(tp)
```
trivial idea... we could have created a common builder with all fields except pubSubTopicPartition and used it to create new builder and add only tp to it

sushantmane commented on clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/pubsub/input/PubSubPartitionSplit.java:
```diff
@@ -0,0 +1,225 @@
+      PubSubPosition startPubSubPosition,
+      PubSubPosition endPubSubPosition,
+      long numberOfRecords,
+      int splitIndex,
```
It doesn't need to be

sixpluszero commented on clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/VenicePushJobConstants.java:
```diff
@@ -108,6 +111,79 @@ private VenicePushJobConstants() {
+   * The split type determines how input splits are generated for processing Kafka records.
+   * Supported values include {@code SplitType} supported by the system.
+   */
+  public static final String KAFKA_INPUT_SPLIT_TYPE = "kafka.input.split.type";
```
What I understand is this is not Kafka specific, this KIF will be PubSubInputFormat and even you add the new class SparkPubSubInputFormat
I understand it is just naming and we have a huge amount of naming dependency for Kafka in OSS code, but what do you think about using the PubSub prefix for newly added configs (for old one maybe not today)?

sushantmane commented:
> can this pr be broken into smaller pieces? there are multiple tasks mentioned in the description - 1) offset to position change 2) Introduce per-consumer monotonic 3) Refactor and abstract previously duplicated record reading and split calculation. can these be separated ?

Extracted: https://github.com/linkedin/venice/pull/2045. Thanks! Will update this PR

sushantmane commented:
There is code duplication that I want to remove (`KafkaInputRecordReader` & `SparkPubSubInputPartitionReader`) but I'll do that in follow up PR

sushantmane commented on internal/venice-common/src/main/java/com/linkedin/venice/pubsub/api/EarliestPosition.java:
```diff
@@ -76,6 +76,6 @@ public int hashCode() {
   @Override
   public long getNumericOffset() {
-    return -1;
+    return -1L;
```
This is not undefined value. Also, it is not supposed to be used via constant.

arjun4084346 commented on clients/venice-push-job/src/main/java/com/linkedin/venice/vpj/VenicePushJobConstants.java:
```diff
@@ -112,18 +112,20 @@ private VenicePushJobConstants() {
-  public static final long DEFAULT_KAFKA_INPUT_MAX_RECORDS_PER_MAPPER = 5000000L;
-
-  public static final String KIF_RECORD_READER_KAFKA_CONFIG_PREFIX = "kif.record.reader.kafka.";
+  public static final long DEFAULT_PUBSUB_INPUT_MAX_RECORDS_PER_MAPPER = 5000000L;
```
* nit
you can write it as `5_000_000` more readable and standard practice in Java

sushantmane replied:
Done

sushantmane commented:
Thanks a lot, @sixpluszero & @arjun4084346!
